wallet-client/WalletClient.hs

```diff
@@ -0,0 +1,223 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module WalletClient where
+
+import           Data.Void (Void)
+import           Data.ByteString.Lazy (ByteString)
+import           Data.Proxy (Proxy (..))
+
+import qualified Codec.Serialise as Serialise (encode, decode)
+import           Network.Socket as Socket
+
+import           Control.Monad.Class.MonadST
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadThrow
+import           Control.Monad.Class.MonadTimer
+import           Control.Tracer
+
+import           Ouroboros.Consensus.Demo
+import           Ouroboros.Consensus.Demo.Run
+import           Ouroboros.Consensus.Mempool
+import           Ouroboros.Consensus.NodeId
+import           Ouroboros.Consensus.Protocol.Abstract (NodeConfig)
+import           Ouroboros.Consensus.Block (BlockProtocol)
+import           Ouroboros.Consensus.Util.Condense (Condense (..))
+
+import           Network.TypedProtocol.Codec
+import           Network.TypedProtocol.Driver
+import           Ouroboros.Network.Block (Point)
+import qualified Ouroboros.Network.Block as Block
+import           Ouroboros.Network.Chain (genesisPoint)
+import           Ouroboros.Network.Codec
+import           Ouroboros.Network.Mux.Interface
+import           Ouroboros.Network.Protocol.LocalTxSubmission.Type
+import           Ouroboros.Network.Protocol.LocalTxSubmission.Client
+import           Ouroboros.Network.Protocol.LocalTxSubmission.Codec
+import           Ouroboros.Network.Protocol.ChainSync.Type
+import           Ouroboros.Network.Protocol.ChainSync.Client
+import           Ouroboros.Network.Protocol.ChainSync.Codec
+import           Ouroboros.Network.NodeToClient
+
+runWalletClient :: forall blk.
+                   RunDemo blk
+                => DemoProtocol blk
+                -> CoreNodeId
+                -> NumCoreNodes
+                -> Tracer IO String
+                -> IO ()
+runWalletClient ptcl nid numCoreNodes tracer = do
+
+    let ProtocolInfo{pInfoConfig} =
+          protocolInfo numCoreNodes
+                       nid
+                       ptcl
+
+        addr = localSocketAddrInfo (localSocketFilePath nid)
+
+    connectTo
+      (muxLocalInitiatorNetworkApplication (Proxy :: Proxy blk) tracer pInfoConfig)
+      Nothing
+      addr
+
+muxLocalInitiatorNetworkApplication
+  :: forall blk m.
+     (RunDemo blk, MonadST m, MonadThrow m, MonadTimer m)
+  -- TODO: the need of a 'Proxy' is an evidence that blk type is not really
+  -- needed here.  The wallet client should use some concrete type of block
+  -- from 'cardano-chain'.  This should remove the dependency of this module
+  -- from 'ouroboros-consensus'.
```

Comment:
In this type signature, `blk` only appears in `BlockProtocol`, which is a non-injective type family. I think you could remove the proxy and flip on `AllowAmbiguousTypes` instead. It did this in https://github.com/input-output-hk/ouroboros-network/pull/635 but @dcoutts didn't like it.

Reply:
Really though I think you should just leave the proxy. It's one of the fee options that we must pay for using type families so liberally, as ouroboros-consensus does.